### ISSUE TITLE
Image upload - Post ID or guid required

### ISSUE THREAD

**Issue by GoogleCodeExporter** [viktorthang/wordpress-custom-content-type-manager#404]
```
What steps will reproduce the problem? (Be specific enough so that someone
else can make the plugin fail for themselves. Remember: if the issue can't
be fixed if it can't be reproduced!)
1. Create new content type
2. Create new custom field (image) and link to content type
3. Create new content based on the content type
4. Use the "Upload" button to add an image (either browse locally or from media 
library)
5. And then click "Insert into post"

What is the expected output? What do you see instead?
Error message: "Post ID or guid required."
This seems simmilar to Issue 367 which is reported as fixed.
This is only a problem with the upload-button. The other button (which I named 
"Choose image") works as it should.



Does the problem continue if you disable all other plugins? (Conflicts
between plugins are easily the #1 cause of errors, so it's important to
isolate the issue):
Yes, disabled all plugins. Problem continues.


Please copy and paste the system information that appears in a yellow
textarea when you click on "Report a Bug" on any CCTM admin screen (this
includes the version of the
plugin, the version of PHP, the version of MySQL, a list of other active
plugins etc.):

*SYSTEM INFO* 
------------------------ 
Plugin Version: 0.9.6-pl
WordPress Version: 3.4.1
PHP Version: 5.3.14
MySQL Version: 5.5.24-cll
Server OS: Linux
------------------------ 
ACTIVE PLUGINS: 
------------------------ 
CURRENT THEME: xxxxx v. http://www.xxxxxx.xx

```

Original issue reported on code.google.com by `to...@spindelvev.net` on 8 Aug 2012 at 9:48

Attachments:
- [bug.png](https://storage.googleapis.com/google-code-attachments/wordpress-custom-content-type-manager/issue-404/comment-0/bug.png)

**Comment by GoogleCodeExporter** [comment]
Original comment by `ever...@fireproofsocks.com` on 8 Aug 2012 at 3:14
- Changed state: **InReview**

**Comment by GoogleCodeExporter** [comment]
```
I can't reproduce this on my local testing server.  The server where this is 
occurring is doing some strange things... when I edit a Javascript file 
(js/relation.js), the changes stick (I can close the file and reopen in via 
FTP, and the changes remain), but when I view the file in a browser, the 
changes are NOT there (yes, even after clearing my browser cache, and even when 
viewing the same file in different browsers).  That makes it nearly impossible 
to test this because I can't edit the javascript to log what it is doing.  That 
could be the entire problem since the upload button (i.e. the cctm_upload() 
function inside of js/relation.js) relies on dynamically redefining an existing 
JS function (that's how I override the default WP behavior).

Possibly related: when viewing the Javascript file in Firebug, we get this 
interesting message:

"The character encoding of the plain text document was not declared. The 
document will render with garbled text in some browser configurations if the 
document contains characters from outside the US-ASCII range. The character 
encoding of the file needs to be declared in the transfer protocol or file 
needs to use a byte order mark as an encoding signature."

Note that I get the same message on my local dev server -- could be this bug in 
Firefox 14.0.1: 
https://www.labkey.org/issues/home/Developer/issues/details.view?issueId=15652
```

Original comment by `ever...@fireproofsocks.com` on 10 Aug 2012 at 8:08
- Changed state: **InProgress**

**Comment by GoogleCodeExporter** [comment]
```
See this screenshot: https://skitch.com/fireproofsocks/em96h/cctm-js-caching

I could not reproduce this behavior with other JS files (??!?) -- I tried 
creating a test.js file and editing it and viewing it in a browser, and those 
changes appeared normally.
```

Original comment by `ever...@fireproofsocks.com` on 10 Aug 2012 at 8:16

**Comment by GoogleCodeExporter** [comment]
```
Ah, got the JS thing sorted -- red herring.  Testing the ajax controllers now.
```

Original comment by `ever...@fireproofsocks.com` on 10 Aug 2012 at 8:48

**Comment by GoogleCodeExporter** [comment]
```
Interesting... on my local dev server, the Upload javascript captures something 
like the following: 

http://mysite.com/wp-content/uploads/2012/08/my-image.jpg

But on the problematic server, the link comes through something like this:

http://mysite.com/?attachment_id=412

So the identifier is not being recognized in the 
ajax-controllers/get_selected_posts.php file.
```

Original comment by `ever...@fireproofsocks.com` on 10 Aug 2012 at 8:57

**Comment by GoogleCodeExporter** [comment]
```
Committed revision 584058.  Will go live with 0.9.6.1.
```

Original comment by `ever...@fireproofsocks.com` on 10 Aug 2012 at 9:14
- Changed state: **Fixed**
- Added labels: **Milestone-Release0.9.7**

**Comment by GoogleCodeExporter** [comment]
```
This problem (“Post ID or guid required”) continued for me in 0.9.6.1. To 
replicate, click "upload" for an image custom field, then set the "Link URL" 
field to none/blank. Note that "Link URL" may be blank by default if 
‘image_default_link_type’ is blank, instead of set to "file" or "post" 
(without quotes) (see admin settings at yoursite.com/wp-admin/options.php).

A non-lasting but coding-free solution is to put a value in the “Link URL” 
field in the media uploader window (e.g., by clicking “File URL” or 
“Attachment Post URL”). Another note: Unless ‘image_default_link_type’ 
is set (see above), I think the uploader will default to use whichever “Link 
URL” setting was most recently used. 

It might also be helpful to set ‘image_default_link_type’ (see admin 
settings at yoursite.com/wp-admin/options.php) to "file" or "post" (without 
quotes) instead of leaving it blank, but a user could still remove the "Link 
URL" in the media upload window, which would give the “Post ID or guid 
required” error message.

I'm hoping for feedback on whether this next solution is appropriate. I think 
the best fix would be to adjust ‘/js/relation.js’. As I write this, its 
‘cctm_upload’ function has this code early on: 

    var matches = html.match(/href=['|"](.*?)['|"]/); 

It believe it is populating a variable ‘matches’ with the guid for the 
image. It gets the guid from the href attribute of the anchor tag that wraps 
image tag. However, if "Link URL" was set to none in the media upload window, 
there will be no anchor wrapper and thus no href, so the guid will not be 
passed from relation.js to the php script at 
‘ajax-controllers/get_selected_psts.php.’ Receiving no guid, the latter 
(php) script returns the error “Post ID or guid required”. 

Perhaps the image tag is more reliably present than the anchor tag, and 
relation.js could use the image tag's src attribute to get the guid. So, the 
fix would be to replace "href" with "src" in the line of code above (from the 
‘cctm_upload’ function), like this:

        var matches = html.match(/src=['|"](.*?)['|"]/);

It's working for me so far, but I haven't tested it much. Ideas?
```

Original comment by `philnels...@gmail.com` on 24 Oct 2012 at 8:29

**Comment by GoogleCodeExporter** [comment]
```
I gotta get the code to field all the whacko stuff WP throws at it.  
```

Original comment by `ever...@fireproofsocks.com` on 24 Oct 2012 at 9:07
- Changed state: **ReOpened**

**Comment by GoogleCodeExporter** [comment]
```
This is a bit of a quandary: the WP code I'm showing here is more of a courtesy 
-- it's poorly written, and clearly not intended to be modular, but it is 
handy.  However, if you remove the link or you choose a different size for your 
image, it becomes increasingly difficult to navigate back to the unique post ID 
(i.e. attachment ID) from which the image came.  

I'm working to increase the error messaging here -- it's not possible to fix 
all of WP's warts in this case.
```

Original comment by `ever...@fireproofsocks.com` on 19 Nov 2012 at 5:43

**Comment by GoogleCodeExporter** [comment]
```
Like you said, if a user removes the link, the <a> tag's href will no longer be 
available to provide the guid. Could you switch to pulling the guid from the 
<img> tag's src attribute instead? It would be present whether or not a link is 
applied to the image. So, you could replace 'href' with 'src' in 
/js/relation.js, as I mentioned above.

I haven't tested what happens to the src/guid when the image size is changed.
```

Original comment by `philnels...@gmail.com` on 19 Nov 2012 at 6:26

**Comment by GoogleCodeExporter** [comment]
```
Yes, I can use the src instead of the href, but it only punts the problem: if 
you use the src, that is more likely to uniquely identify the image, but if the 
user selects any other size other than "Full Sized", the src attribute is not 
stored in the database. 

For example, the full-sized image is stored verbatim in the wp_posts.guid 
field, e.g.

http://mysite.com/wp-content/uploads/2012/06/IMG_0448.jpg

(i.e. SELECT ID from wp_posts WHERE 
guid='http://mysite.com/wp-content/uploads/2012/06/IMG_0448.jpg' will return 
exactly 1 row so we can determine the post id).

But if you click on "Thumbnail" or any other image size, the URL that WP uses 
is something like this:

http://mysite.com/wp-content/uploads/2012/06/IMG_0448-150x150.jpg

That value does NOT appear in the database... so I'd have to do some regex 
manipulation on the URL to guess what the original image was and thus get its 
post ID.  It's tricky -- it'd be nice if WP had built this to be modular and 
normalized in the first place because patching leaky abstractions takes a lot 
of time and it's highly vulnerable to breaking if and when WP changes their 
conventions.

See also Issue #403 -- it's a similar problem.  It may be cleaner to remove 
this link entirely.... it's just a tarbaby.
```

Original comment by `ever...@fireproofsocks.com` on 19 Nov 2012 at 6:58

**Comment by GoogleCodeExporter** [comment]
```
Ah, I see. So it wouldn't be a straight-shot to use src as the guid. 

This stuff is still unfamiliar to me, so I hope I'm not wasting your time with 
more reactions. Don't feel like you even need to respond.

I uncommented your 'alert(html)' code in relation.js ('thumbnail' and no link). 
The pop-up:

        <img src="http://mysite.com/wp-content/uploads/2012/11/test-image-150x100.gif" alt="" title="test-image" 
        width="150" height="100" class="alignnone size-thumbnail wp-image-1881" />

If those class selectors will be present/consistent across all WP 
configurations, the wp-image-1881 may have the most potential. 1881 is the ID 
value of the newly uploaded post in wp_posts. So, maybe it could facilitate a 
SQL query like SELECT GUID from wp_posts WHERE ID=1881.

I see that same ID number appear in the source for the "Insert Into Post" 
submit button, but I'm not up to speed enough to know whether/how your script 
could grab the ID from there.

        <input type="submit" name="send[1881]" id="send[1881]" class="button" value="Insert into Post"

If neither of the above work, and you end up having to do the regex 
manipulations to guess at the original image URL, maybe using the 
'size-thumbnail' class selector in combination with 
get_option('thumbnail_size_w') and get_option('thumbnail_size_h') to find the 
'-150x100' to remove from the src in guessing the guid.

Ok, one more idea. The image_size_input_fields() function (line 712 at 
http://core.trac.wordpress.org/browser/tags/3.4.2/wp-admin/includes/media.php) 
creates the markup listing image sizes on the image-upload modal window. Maybe 
its techniques for determining available image sizes (needed for your regex) 
could be borrowed. 

Sorry, last one. I didn't read it, but maybe there is something useful here: 
http://wp.tutsplus.com/tutorials/creative-coding/how-to-integrate-the-wordpress-
media-uploader-in-theme-and-plugin-options/
```

Original comment by `philnels...@gmail.com` on 20 Nov 2012 at 4:35

**Comment by GoogleCodeExporter** [comment]
```
Ah, yes, grabbing the ID from the CSS class is a good idea... and that looks 
like a good link.  Will check them out.  This would solve the problem I was 
having with Issue #403 too.  Thank you for your valuable input!
```

Original comment by `ever...@fireproofsocks.com` on 20 Nov 2012 at 5:02
- Changed state: **InProgress**

**Comment by GoogleCodeExporter** [comment]
```
Again, thanks. 

You mentioned above "...the WP code I'm showing here is more of a courtesy." I 
totally agree.

I think your whole plugin is a courtesy, actually... a huge gift. Seriously. I 
mean, I can't imagine the endless hours  you've invested, creating so much 
value, then just granted everyone access.
```

Original comment by `philnels...@gmail.com` on 20 Nov 2012 at 2:09

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Hi guys i was using the CCTM plugin and was having trobule with the uploader 
too. I get that annoying message too. I found that the problem was in 
relation.js, where fill "matches" var. I replace the line that fills the var 
for this:

var matches;

if (html.indexOf("src") != -1){
    matches = html.match(/src=['|"](.*?)['|"]/);
}else if (html.indexOf("href")) {
    matches = html.match(/href=['|"](.*?)['|"]/);
}

Maybe i had other problem and just this work for me, hope it can help anybody.

See you!
```

Original comment by `f...@ingamana.com` on 19 Dec 2012 at 6:58

**Comment by GoogleCodeExporter** [comment]
```
Thanks for sharing your code.
```

Original comment by `ever...@fireproofsocks.com` on 19 Dec 2012 at 7:27

**Comment by GoogleCodeExporter** [comment]
```
This isn't going to be a 100% clean fix since the WP uploader itself is not 
intended for use in the way that the CCTM is using it, but the updates have 
been applied and will be released in 0.9.8.  Committed revision 646431 or 
before...
```

Original comment by `ever...@fireproofsocks.com` on 31 Dec 2012 at 4:56

**Comment by GoogleCodeExporter** [comment]
Original comment by `ever...@fireproofsocks.com` on 31 Dec 2012 at 4:58
- Changed state: **Fixed**
